### PR TITLE
move reconciliation out of transaction

### PR DESCRIPTION
**Pull request by jmtroth0**
https://github.com/dimagi/commcare-hq/pull/22341#issuecomment-437279281 yeah, this seems to work as well, although i'm not enthused by the call to `toggles.SORT_OUT_OF_ORDER_FORM_SUBMISSIONS_SQL` for each case. can i assume here that each case will have the same domain? or is it worth checking what domains exist and what have it set to true beforehand?

@snopoke @millerdev buddy: @orangejenny

### CODE REVIEW

**Review by millerdev — APPROVED** [review]
> can i assume here that each case will have the same domain? or is it worth checking what domains exist and what have it set to true beforehand?

I think it would be an indication of a bug if you found multiple domains. Although maybe not if case sharing is turned on? @snopoke, do you know? Should probably check either way. Something like this should work since there should be very few domains (1 in most cases).

```py
domains = {case.domain for case in cases}
sort_submissions = {
    d for d in domains
    if toggles.SORT_OUT_OF_ORDER_FORM_SUBMISSIONS_SQL.enabled(
        d, toggles.NAMESPACE_DOMAIN)
}
for case in cases:
    if case.domain in sort_submissions:
        # do sorting...
```

**Review by snopoke — COMMENTED** [review]
There can only ever be one domain here since all the cases being saved will come from a single form.

**Review by millerdev — APPROVED** [review]

File: corehq/form_processor/backends/sql/processor.py [hunk 1]
```diff
@@ -145,16 +145,20 @@ def save_processed_models(cls, processed_forms, cases=None, stock_result=None, p
...
                 ledgers_to_save = stock_result.models_to_save
                 LedgerAccessorSQL.save_ledger_values(ledgers_to_save, stock_result)
 
+        if cases:
+            sort_submissions = toggles.SORT_OUT_OF_ORDER_FORM_SUBMISSIONS_SQL.enabled(
+                processed_forms.submitted.domain, toggles.NAMESPACE_DOMAIN)
+            if sort_submissions:
+                for case in cases:
+                    SqlCaseUpdateStrategy(case).reconcile_transactions_if_necessary()
+                    CaseAccessorSQL.save_case(case)
```
Comment (snopoke) [inline corehq/form_processor/backends/sql/processor.py]: Surely we only need to save the case if it actually changed right?

Comment (jmtroth0) [reply corehq/form_processor/backends/sql/processor.py]: https://github.com/dimagi/commcare-hq/pull/22447

Comment (jmtroth0) [reply corehq/form_processor/backends/sql/processor.py]: oops

**Pull request merged**